src/runtime/file_utils.cc

```diff
@@ -64,12 +66,14 @@ void FunctionInfo::Save(dmlc::Stream* writer) const {
   writer->Write(name);
   writer->Write(arg_types);
   writer->Write(thread_axis_tags);
+  writer->Write(use_dyn_shared_memory);
 }
 
 bool FunctionInfo::Load(dmlc::Stream* reader) {
   if (!reader->Read(&name)) return false;
   if (!reader->Read(&arg_types)) return false;
   if (!reader->Read(&thread_axis_tags)) return false;
+  if (!reader->Read(&use_dyn_shared_memory)) return false;
```

Comment:
@tqchen I didn't find a way to optionally read this attribute for the binary reader.

Reply:
I think. let us think about another way to do backward compact then. 

Since binary backward compatibility is more important to us here.

Here is one solution that also helps future backward compatibility.

Let us rename `thread_axis_tags => launch_param_tags`, and use `dyn_shared_memory` as a tag for the launching parameter of dynamic shared memory. We can then change the `ThreadAxisConfig => LaunchParamConfig`.

Reply:
Let us try the ^ change and do a test to make sure we can load back old cuda binaries

Reply:
If we do the rename `thread_axis_tags => launch_param_tags`, the JSON reader would break  backward compat at `helper.DeclareField("launch_param_tags", &launch_param_tags);` .

Or shall we do the rename but keep the old attribute name for JSON reader/writer? e.g.
`helper.DeclareField("thread_axis_tags", &launch_param_tags);`.

Reply:
Went with the latter solution, and confirmed that an old saved module can be loaded via both json and binary.

Reply:
We can remove the thread_axis_tags later